### ISSUE TITLE
Re-Think Versioning

### ISSUE THREAD

**Issue by jGleitz** [robstoll/atrium-roadmap#61]
The main repository has [a rough roadmap](https://github.com/robstoll/atrium#roadmap) that assigns features to version numbers. This roadmap does not follow semantic versioning, because it ties non-breaking changes _and_ breaking changes to changes in the `MAJOR` version.

Proposal: We only plan which _breaking_ changes will happen with which major versions. All other features will just be released when they are ready. We use branches for the breaking changes that should not be released yet.

I propose these changes because:

 a) We should not block new features by not releasing breaking changes (unless those features depend on the breaking change, of course).
 b) I feel that the current versioning discourages PRs. JSON assertions are currently planned for `0.11.0`, for example. This gives me the feeling that a PR adding some simple JSON assertions would not be welcome at this point.

Additionally, we could stop doing a roadmap in atrium’s README and use milestones in this project instead. (We would only have milestones for breaking changes, though). I think this would make it easier to plan. We could still link to the milestone view from the README.

This ticket is inspired by the discussion in #60

**Comment by robstoll** [comment]
I see that the roadmap can be misleading. It has to be understood as a rough plan of us core contributors where things can of course change such as a feature might come earlier if a contributor jumps in. 

> We use branches for the breaking changes that should not be released yet.

This one I don't see as practical for me as I am breaking things regularly (we are still in 0.x) and I don't want the extra job of cherry-picking features on a separate branch.

> a) We should not block new features by not releasing breaking changes (unless those features depend on the breaking change, of course).

I agree. Yet, the current situation is a bit special maybe because all new features are based on `Expect`. And the transition from `Assert` to `Expect` should be complete IMO otherwise it will be painful experience to migrate. 

> b)  I feel that the current versioning discourages PRs. JSON assertions are currently planned for 0.11.0, for example. This gives me the feeling that a PR adding some simple JSON assertions would not be welcome at this point.

this goes hand in hand with 

> We only plan which breaking changes will happen with which major versions. All other features will just be released when they are ready.

and

> Additionally, we could stop doing a roadmap in atrium’s README and use milestones in this project instead. I think this would make it easier to plan. We could still link to the milestone view from the README.

We have already milestones, I wouldn't use it for breaking changes only though. 
Good with me to change the roadmap section in the readme to only mention planned breaking changes. Especially as we have the atrium-roadmap now where one can look what ideas are floating around. Do you have time to create a PR?

So far so good, I don't see a proposition to change the versioning though. So maybe I am missing something between the lines.

**Comment by robstoll** [comment]
Some more information about the current versioning. I basically try to following semantic versioning for the API modules. Hence no breaking changes, deprecated functionality stays in there until a next major version.

We already discussed that we want to kick deprecated things out with 0.10.0. 
Thus IMO we could consider to release 0.10.0 as 1.0.0 and use normal semantic versioning afterwards. Which means we will probably regularly release a major version as I have the feeling the project is still not at a stage where I see it less likely that we do huge refactorings and with them breaking changes.

However, what I do not want is creating 1.1.0, 1.2.0 etc. and work in parallel on a 2.0.0

**Comment by jGleitz** [comment]
> This one I don't see as practical for me as I am breaking things regularly (we are still in 0.x) and I don't want the extra job of cherry-picking features on a separate branch.

> However, what I do not want is creating 1.1.0, 1.2.0 etc. and work in parallel on a 2.0.0

I agree that branching might create some overhead. However, I still hope we can switch to that model. Because currently, breaking changes block the release of features that are not directly connected to the breaking change. 

Let’s take the `Assert` / `Expect` example. I have not had a look at what exactly needs to be changed, but I think this is how it _should_ be:

 * New `Expect` gets merged into `master`
 * New assertions can be merged into `master`, but they can only be used by users that have switched to `Expect`
 * We prepare the removal of `Assert` on a separate branch which we merge into `master` once it’s ready.

So, in general, a new feature either targets `master` or a breaking-release-branch. In the first case, it should be easy to merge `master` into the breaking-release-branch. If merging would be complicated, _the feature should not target `master`_ but the breaking-release-branch.

The last “stable” release of atrium was 9 months ago. Any work done after that has not been released to most users, even though they could profit from some of the changes. I think this is a bad thing. It is the main motivation why I propose the “master is always releasable” convention.

In conclusion: Branching should enable us to publish changes as soon as possible. If we do it right, it should create little overhead because merging should be easy most of the time.

<hr>

> I wouldn't use it for breaking changes only though.

I think it does not make sense to assign a milestone to a feature request if it doesn’t depend on a breaking change. Because in that case, the feature should be released as soon as possible and we cannot foresee the version number.

We can indicate that feature requests depend on breaking changes by using milestones, though.

<hr>

> Good with me to change the roadmap section in the readme to only mention planned breaking changes. Especially as we have the atrium-roadmap now where one can look what ideas are floating around. Do you have time to create a PR?

I am happy to create a PR. However, can we remove the roadmap altogether? I feel it is one more thing to keep consistent and as high potential to be out of date. We could tag issues with `breaking` instead and link to a view that shows all these issues. This way, users will see which breaking changes are planned for which milestones. And we have one less thing to keep up-to-date.

<hr>

> Thus IMO we could consider to release 0.10.0 as 1.0.0 and use normal semantic versioning afterwards. Which means we will probably regularly release a major version as I have the feeling the project is still not at a stage where I see it less likely that we do huge refactorings and with them breaking changes.

Major version changes are not a bad thing in and of themselves. What speaks in favour of this plan is that atrium is already pretty useful and therefore might deserve a “proper” release.

**Comment by robstoll** [comment]
> However, I still hope we can switch to that model. 

OK, I can try it out. But if it bothers me too much, then I'll drop the idea again.

> Let’s take the Assert / Expect example [...]

I don't like half baked changes in general but it is of course debatable when it is half backed. 
For the `Expect` story it would have meant that for instance you added `Expect<Path>` but there is not even `Expect<T>.toBe` implemented. And therefore I would not have wanted to promote it in a release.
In this sense I am not particularly a fan of  `release always` but I would like to embrace the `release often` mantra. The last stable release is too long ago, that's also true, I actually though I could release it February last year but then started working on `Expect` .

Thinking about it. We could already release 0.9.0 today the only thing missing in 0.9.0 is the infix API and we could easily release it with the next version. What do you think?

Following the problems I encountered so far with rolling releases in combination with semantic versioning.
- migration from one version to the next is painful (not always of course): missing migration guides,  using new features does only work so so as the new feature set is not complete
- documentation/KDoc: there are too many versions and a search engine brings you to the wrong version first
- 
- new major versions are released too often, so that if you use the library in a company which has restrictions on how fast you are allowed to update major versions, has to update several major versions at one in the end.
- new functionality has to be introduced as experimental to circumvent semantic versioning

I would like to give good migration guidelines so that switching from one version to the next is not painful. Also, I am following semantic versioning closely, changing a signature of a method (e.g. add a default value) is a breaking change to me as it breaks binary compatibility. Therefore I usually let new features settle for a while because many times I find errors where a fix requires a breaking change. 

A conclusion is given at the end of this comment

> I think it does not make sense to assign a milestone to a feature request if it doesn’t depend on a breaking change. 

Currently it is more like the following. An issue number is not assigned to a milestone unless I already know that it should be included in the release (like breaking changes). Once an issue is resolved it gets the version of the next milestone as the feature will be released with it. I am using the milestone as detailed release-notes. Maybe this would change with #60 
However, I suggest we use the Milestones in atrium-roadmap the way you describe. Only assign issues to milestones which include a breaking change. 

> I am happy to create a PR. However, can we remove the roadmap altogether? 

Good with me if we use the milestones/project in atrium-roadmap for the purpose of communicating breaking changes. Therefore, please create corresponding issues and milestones  in atrium-roadmap

> Major version changes are not a bad thing in and of themselves. What speaks in favour of this plan is that atrium is already pretty useful and therefore might deserve a “proper” release.

OK, let's go in this direction

# Conclusion
Please state if you agree to the following plan:
1. release 0.9.0 without the infix API immediately
2. release the infix API as 0.10.0
3. release any new features as 0.11.0, 0.12.0 etc. 
4. introduce the breaking changes which are currently foreseen for 0.10.0 (get rid of ServiceLoader etc.) in a separate branch
5. I'll merge the new domain back into master as soon as it is stable
6. you are porting new features since 0.10.0 to the new domain
7. I'll continue with other work on a separate branch and release 1.0.0 as soon as stable.

**Comment by jGleitz** [comment]
I agree with the plan given in your conclusion.

I also agree that we should take migration paths seriously. It’s very impressive how well you care for users of atrium. We should keep that up and do nothing that impacts it negatively.

I hope that the branching model will allow us to prepare releases as carefully as needed. I hope also that we will collect breaking changes in branches and avoid releasing major versions too often. However, the branching model should also allow us to release new features while we do the preparation. I.e. preparing breaking changes will no longer block features.

If porting changes between the branches is what I can do to help atrium gain a faster release cadence, I am happy to do it.

**Comment by robstoll** [comment]
> If porting changes between the branches is what I can do to help atrium gain a faster release cadence, I am happy to do it.

How about you help out transforming the current infix API to the new API?
https://github.com/robstoll/atrium/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+new+infix

**Comment by jGleitz** [comment]
I see that you have already modified the readme. Since we seem to agree to release `1.0.0` soon and to switch to a branching model (i.e. breaking changes are prepared on separate branches), I think this issue can be closed.
